### DISCUSSION TITLE
TypeForm + TypeVar

### DISCUSSION THREAD

**Discussion by superlopuh** [microsoft/pyright#9462] [Q&A] [answered]
I have a question about passing `TypeVar`s to functions that expect `TypeForm`s. Is my understanding of the proposal correct that this is one of the planned supported use-cases? Pyright currently seems not to support it, so I thought I'd ask here whether this is intentional:

``` python
from typing_extensions import TypeForm, TypeVar
class A: ...
B = TypeVar("B", bound=A)
def func(tf: TypeForm[A]) -> TypeForm[A]: ...
def this_works(val: B) -> B:
    # B is a valid spelling of a type in this context
    val0: B = val
    # Argument of type "TypeVar" cannot be assigned to parameter "tf" of type "TypeForm[A]" in function "func" "TypeVar" is not assignable to "TypeForm[A]" (reportArgumentType)
    val1 = func(B)
    return val0
```

**Comment by erictraut** [comment, marked as answer]
You need to set `"enableExperimentalFeatures"` to true to enable `TypeForm` support. Once you do this, the above code type checks without error.

Code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmB9ApgDxllAZyTCKiQgTBBigBVEsAxaiAGnsYDUBDEAKADGAGx6FCUAIIAuKADoF-AEJQAvJwRZeIABQAiJXo4AjMAFcUAE1WSAlP0tZgUYBcE6YwWQ00sQEAG1JAF1bKABaAD4NZlYg4NkFOQcnWAALJEIMAHdqAGtCHQA3HmFZJTCoqCVpfih6qABiavIJHigS4SRLKEJNYS70KDBndvhNchR0zKhBUnw8OobOgAZytQ7SpfrmyRA0MwgCWhHYRig9Hy0%2BPVmeFBQwWmMsKDFiNBQsHpgwKAQ%2BDwjvgQBdPLdTuNXpdGH5AiFbqgXG4YCQpnpXChBLcYZptIiJI9aO8kJ8eMZhK9fhcrnD4rcdCAsFQaHsDkcUDArvYGpthABGDaY9wVbZQJkwMwgKarfhAA)

**Reply by superlopuh** [reply]
Amazing, thank you

**Reply by superlopuh** [reply]
Am I right in understanding that the current shipping version of typing_extensions does not export TypeForm, and that the way to use it via `if TYPE_CHECKING:` and in quotes?

**Reply by erictraut** [reply]
I'm not sure if the latest version of `typing_extensions` has support for this. The PEP is still in draft form, so the design may still change. I recommend that you don't use it in any real code. It's implemented in pyright for experimentation so people can provide feedback on the PEP.
